File: clients/vscode/test/status.test.ts
import * as assert from 'node:assert/strict';
import test from 'node:test';

import {
  GO_TO_SOURCE_COMMAND,
  bar,
  icon,
  isFileHealth,
  renderStatus,
  type FileHealth,
} from '../src/status.js';

const healthy: FileHealth = {
  uri: 'file:///a.rs',
  quality: 87.4,
  grade: 'excellent',
  functions: 4,
  below: 0,
  bands: { excellent: 3, good: 1, fair: 0, poor: 0 },
  worst: [
    {
      name: 'add',
      quality: 80.2,
      grade: 'excellent',
      line: 12,
      weakestPillar: 'interface',
      weakestMetric: 'parameters',
    },
  ],
};

test('a bar fills in proportion to the score', () => {
  assert.equal(bar(0), '░░░░░░░░░░');
  assert.equal(bar(100), '██████████');
  assert.equal(bar(50), '█████░░░░░');
  assert.equal(bar(-20), '░░░░░░░░░░', 'a nonsense score does not overflow the bar');
  assert.equal(bar(150), '██████████');
});

test('a bar resolves a fraction of a cell', () => {
  // Eight steps within every cell, so two scores a percent apart do not
  // draw the same bar.
  assert.equal(bar(6.25), '▋░░░░░░░░░');
  assert.equal(bar(55), '█████▌░░░░');
  assert.notEqual(bar(51), bar(54));
  // However it divides, a bar is always the same width.
  for (let step = 0; step <= 200; step += 1) {
    assert.equal([...bar(step / 2)].length, 10, `bar(${step / 2})`);
  }
  assert.equal([...bar(Number.NaN)].length, 10, 'a NaN score still draws a bar');
});

test('the bar shows the band and the rounded score', () => {
  const rendered = renderStatus(healthy);
  assert.equal(rendered.text, '$(pass) 87%');
  assert.equal(rendered.warning, false);
});

test('functions below the threshold are counted in the bar itself', () => {
  const rendered = renderStatus({ ...healthy, below: 3 });
  assert.match(rendered.text, /\$\(alert\)3$/);
});

test('a bad file is coloured and iconed differently', () => {
  for (const [grade, expected, warning] of [
    ['poor', '$(error)', true],
    ['fair', '$(warning)', true],
    ['good', '$(check)', false],
  ] as const) {
    const rendered = renderStatus({ ...healthy, grade });
    assert.equal(icon(grade), expected);
    assert.ok(rendered.text.startsWith(expected), rendered.text);
    assert.equal(rendered.warning, warning);
  }
});

test('the tooltip carries the spread across bands', () => {
  const { tooltip } = renderStatus(healthy);
  assert.match(tooltip, /\*\*87%\*\*/);
  assert.match(tooltip, /4 functions, 0 below the warning threshold/);
  assert.match(tooltip, /\| \$\(pass\) \| excellent \| 3 \|/);
  assert.match(tooltip, /\| \$\(error\) \| poor \| 0 \|/);
});

test('each worst function links to its line', () => {
  const { tooltip } = renderStatus(healthy);
  assert.match(tooltip, /`add`/);
  assert.match(tooltip, /80%/);
  assert.match(tooltip, /weakest interface \(parameters\)/);
  const encoded = encodeURIComponent(JSON.stringify(['file:///a.rs', 12]));
  assert.ok(
    tooltip.includes(`command:${GO_TO_SOURCE_COMMAND}?${encoded}`),
    `no go-to link in:\n${tooltip}`,
  );
});

test('the tooltip offers the actions the status bar cannot', () => {
  const { tooltip } = renderStatus(healthy);
  assert.match(tooltip, /command:workbench\.actions\.view\.problems/);
  assert.match(tooltip, /command:workbench\.action\.openSettings/);
  assert.match(tooltip, /command:lspfAnalysis\.restartServer/);
});

test('no bar is wrapped in a code span', () => {
  // VS Code draws inline code with a background and padding, which would
  // put a gap on either side of every bar.
  const { tooltip } = renderStatus(healthy);
  assert.doesNotMatch(tooltip, /`[█▉▊▋▌▍▎▏░▓▒]/, tooltip);
  assert.doesNotMatch(tooltip, /[█▉▊▋▌▍▎▏░▓▒]`/, tooltip);
});

test('one function is not pluralized', () => {
  const { tooltip } = renderStatus({ ...healthy, functions: 1 });
  assert.match(tooltip, /1 function,/);
});

test('a file with no functions says so instead of showing empty bands', () => {
  const { tooltip } = renderStatus({
    ...healthy,
    functions: 0,
    quality: 100,
    bands: { excellent: 0, good: 0, fair: 0, poor: 0 },
    worst: [],
  });
  assert.match(tooltip, /No functions to analyze/);
  assert.doesNotMatch(tooltip, /Worth opening first/);
});

test('a payload from an unrecognizable server is ignored', () => {
  assert.equal(isFileHealth(healthy), true);
  assert.equal(isFileHealth(undefined), false);
  assert.equal(isFileHealth(null), false);
  assert.equal(isFileHealth('nonsense'), false);
  assert.equal(isFileHealth({ ...healthy, quality: 'high' }), false);
  assert.equal(isFileHealth({ ...healthy, quality: Number.NaN }), false);
  assert.equal(isFileHealth({ uri: 'file:///a.rs' }), false);
});

test('a payload from the earlier three-pillar server is rejected', () => {
  // It sent `worst` as a single object and had no bands at all; rendering
  // it would produce a tooltip full of `undefined`.
  const old = {
    uri: 'file:///a.rs',
    quality: 87.4,
    grade: 'excellent',
    functions: 4,
    below: 0,
    worst: { name: 'add', quality: 80.2, grade: 'excellent', line: 12 },
  };
  assert.equal(isFileHealth(old), false);
});

File: clients/intellij/src/test/kotlin/com/github/meymchen/lspfanalysis/model/StatusTest.kt
package com.github.meymchen.lspfanalysis.model

import com.google.gson.JsonParser
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class StatusTest {

    private fun health(
        quality: Double = 63.4,
        grade: String = "good",
        functions: Int = 7,
        below: Int = 1,
        bands: Bands = Bands(excellent = 4, good = 2, fair = 0, poor = 1),
        worst: List<WorstFunction> = emptyList(),
    ) = FileHealth("file:///src/lib.rs", quality, grade, functions, below, bands, worst)

    @Test
    fun `the widget shows the score and how many want attention`() {
        assertEquals("63%  ⚠1", renderStatus(health()).text)
    }

    @Test
    fun `nothing below the threshold means nothing to flag`() {
        assertEquals("63%", renderStatus(health(below = 0)).text)
    }

    @Test
    fun `only a file bad enough to act on is coloured`() {
        assertFalse(renderStatus(health(grade = "excellent")).warning)
        assertFalse(renderStatus(health(grade = "good")).warning)
        assertTrue(renderStatus(health(grade = "fair")).warning)
        assertTrue(renderStatus(health(grade = "poor")).warning)
    }

    @Test
    fun `the tooltip counts the functions and spells out the bands`() {
        val tooltip = renderStatus(health()).tooltip
        assertTrue(tooltip, tooltip.contains("7 functions, 1 below the warning threshold."))
        assertTrue(tooltip, tooltip.contains("excellent"))
        assertTrue(tooltip, tooltip.contains("poor"))
    }

    @Test
    fun `one function is said in the singular`() {
        val tooltip = renderStatus(health(functions = 1, bands = Bands(0, 1, 0, 0))).tooltip
        assertTrue(tooltip, tooltip.contains("1 function, 1 below the warning threshold."))
    }

    @Test
    fun `a file with nothing to analyze says so instead of counting`() {
        val tooltip = renderStatus(health(functions = 0, below = 0, bands = Bands(0, 0, 0, 0))).tooltip
        assertTrue(tooltip, tooltip.contains("No functions to analyze in this file."))
    }

    @Test
    fun `a popup row names the function and what dragged it`() {
        val worst = WorstFunction(
            name = "tangled",
            quality = 22.1,
            grade = "poor",
            line = 84,
            weakestPillar = "control flow",
            weakestMetric = "cognitive complexity",
        )
        assertEquals(
            "tangled — 22%, weakest control flow (cognitive complexity)",
            worstLabel(worst),
        )
    }

    @Test
    fun `a pillar with one measure names only the pillar`() {
        val worst = WorstFunction("wide", 40.0, "fair", 12, "interface", "")
        assertEquals("wide — 40%, weakest interface", worstLabel(worst))
    }

    private val payload = """
        {
          "uri": "file:///src/lib.rs",
          "quality": 63.4,
          "grade": "good",
          "functions": 7,
          "below": 1,
          "bands": { "excellent": 4, "good": 2, "fair": 0, "poor": 1 },
          "worst": [
            {
              "name": "tangled",
              "quality": 22.1,
              "grade": "poor",
              "line": 84,
              "weakestPillar": "control flow",
              "weakestMetric": "cognitive complexity"
            }
          ]
        }
    """.trimIndent()

    @Test
    fun `the documented notification parses`() {
        val parsed = parseFileHealth(JsonParser.parseString(payload))
        assertNotNull(parsed)
        assertEquals(7, parsed!!.functions)
        assertEquals(4, parsed.bands.excellent)
        assertEquals(84, parsed.worst.single().line)
    }

    @Test
    fun `an unrecognizable notification is ignored rather than rendered`() {
        val cases = listOf(
            "null",
            "\"nonsense\"",
            """{"uri": "file:///a.rs", "quality": "high", "grade": "good", "functions": 1,
               "below": 0, "bands": {"excellent":1,"good":0,"fair":0,"poor":0}, "worst": []}""",
            // A bands object missing one of its four counts.
            """{"uri": "file:///a.rs", "quality": 50.0, "grade": "good", "functions": 1,
               "below": 0, "bands": {"excellent":1,"good":0,"fair":0}, "worst": []}""",
            // A worst entry without the line to jump to.
            """{"uri": "file:///a.rs", "quality": 50.0, "grade": "good", "functions": 1,
               "below": 0, "bands": {"excellent":1,"good":0,"fair":0,"poor":0},
               "worst": [{"name": "a", "quality": 1.0, "grade": "poor"}]}""",
        )
        for (case in cases) {
            assertNull(case, parseFileHealth(JsonParser.parseString(case)))
        }
    }
}
